jaxopt/_src/osqp.py

```diff
@@ -0,0 +1,602 @@
+    z_next = projection_box(z_free, box)
+
+    # line 7: gradient descent on dual variables, with relaxation
+    y_step = tree_sub(z_momentum, z_next)
+    y_next = tree_add(y, tree_scalar_mul(rho_bar, y_step))
+
+    return (x_next, z_next), y_next, x_bar
+
+  def update(self, params, state, params_obj, params_eq, params_ineq):
+    """Perform OSQP step."""
+    # The original problem on variables (x,z) is split into TWO problems
+    # with variables (x, z) and (x_bar, z_bar)
+    # 
+    # (x_bar, z_bar) is NOT part of the state because it is recomputed at each step:
+    #    (x_bar, z_bar) = argmin_{x_bar, z_bar} L(x_bar, z_bar, x, z, y)
+    # with L the augmented Lagrangian
+    # z_bar is always such that A x_bar = z_bar
+    #
+    # x = argmin_x L(x_bar, z_bar, x, z, y)
+    # for equality constraint x = x_bar the dual variable is constant (=0) and can be eliminated
+    #
+    # z = argmin_z L(x_bar, z_bar, x, z, y)
+    # for equality constraint z = z_bar the associated dual variable is y
+    Q    = self.matvec_Q(params_obj[0])
+    c    = params_obj[1]
+    A    = self.matvec_A(params_eq)
+    l, u = params_ineq
+
+    # for active constraints (in particular equality constraints) high stepsize is better
+    rho_bar = state.rho_bar
+    if self.verbose >= 2:
+      print(f"rho_bar={rho_bar}")
+
+    (x, z), y, eq_qp_last_sol = self._admm_step(params, Q, c, A, (l, u), rho_bar, state)
+    if self.verbose >= 3:
+      print(f"x={x}\nz={z}\ny={y}")
+
+    primal_residuals, dual_residuals = self._compute_residuals(Q, c, A, x, z, y)
+    if self.verbose >= 3:
+      print(f"primal_residuals={primal_residuals}, dual_residuals={dual_residuals}")
+
+    rho_bar, params_precond = jax.lax.cond(jnp.mod(state.iter_num, self.stepsize_updates_frequency) == 0,
+      lambda _: self._update_stepsize(rho_bar, state.params_precond, primal_residuals, dual_residuals, Q, c, A, x, y),
+      lambda _: (rho_bar, state.params_precond), operand=None)
+
+    sol = OSQP._get_full_KKT_solution(primal=(x, z), y=y)
+
+    error, status = jax.lax.cond(jnp.mod(state.iter_num, self.termination_check_frequency) == 0,
+      lambda _: self._check_termination_conditions(primal_residuals, dual_residuals,
+                                                   params, sol, Q, c, A, l, u),
+      lambda s: (state.error, s), operand=(state.status))
+
+    state = OSQPState(iter_num=state.iter_num+1,
+                      error=error,
+                      status=status,
+                      primal_residuals=primal_residuals,
+                      dual_residuals=dual_residuals,
+                      rho_bar=rho_bar,
+                      eq_qp_last_sol=eq_qp_last_sol,
+                      params_precond=params_precond)
+    return base.OptStep(params=sol, state=state)
+
+  def run(self, init_params, params_obj, params_eq, params_ineq) -> base.OptStep:
+    """Return primal/dual variables.
+    
+    Args:
+      init_params: initial KKTSolution (can be None).
```

Comment:
Does it need to be a KKTSolution object?

Reply:
What would be the alternative ? For a true warm start we would like to pass dual variables too.

Reply:
I was wondering if you access the attributes or if you treat it as a tuple.

Reply:
I access the attributes for code clarity, but the `KKTSolution` object is itself a tuple and can be build as such, so it should not be too much of a burden for the user to add it.